src/DotNetty.Codecs/Base64/Base64.cs

```diff
@@ -0,0 +1,379 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Codecs.Base64
+{
+    using System;
+    using System.Diagnostics.Contracts;
+    using DotNetty.Buffers;
+
+    public class Base64
+    {
+        const int MAX_LINE_LENGTH = 76;
+        const byte EQUALS_SIGN = (byte)'='; //pad
+        const byte NEW_LINE = (byte)'\n';
+        const sbyte WHITE_SPACE_ENC = -5; // Indicates white space in encoding
+        const sbyte EQUALS_SIGN_ENC = -1; // Indicates equals sign in encoding
+
+        public static IByteBuffer Encode(IByteBuffer src)
+        {
+            return Encode(src, Base64Dialect.STANDARD);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, Base64Dialect dialect)
+        {
+            return Encode(src, src.ReaderIndex, src.ReadableBytes, dialect.breakLinesByDefault, dialect);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, bool breakLines, Base64Dialect dialect)
+        {
+            return Encode(src, src.ReaderIndex, src.ReadableBytes, breakLines, dialect);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect)
+        {
+            return Encode(src, offset, length, breakLines, dialect, src.Allocator);
+        }
+
+        private unsafe static int EncodeUsingPointer(byte* alphabet, IByteBuffer src, IByteBuffer dest, int offset, int length, bool breakLines)
+        {
+            //avoid unnecessary range checking
+            fixed (byte* srcArray = src.Array, d = dest.Array)
+            {
+                byte* destArray = d + dest.ArrayOffset + dest.WriterIndex;
+                int j = 0;
+                int charCount = 0;
+                //commonly, ReaderIndex and ArrayOffset are zero
+                //but you can use SlicedByteBuffer to construct one buffer which ReaderIndex and ArrayOffset are both none zero
+                //for example:
+                //var buff = buffer.Slice(1, 100);
+                //buff.ReadInt();
+                //now ArrayOffset is 1 and ReaderIndex is 4
+                //!!!
+                //and the real offset of the array, is ArrayOfffset + offset
+                //!!!
```

Comment:
this comment seems misplaced. Looks more appropriate in IByteBuffer.ArrayOffset xml doc in some way